infrastructure/code/backend_service.go

```diff
@@ -194,52 +288,38 @@ func (s *SnykCodeHTTPClient) doCall(ctx context.Context,
 			return nil, err
 		}
 	}
+	return req, nil
+}
 
-	// Setting a chosen org name for the request
-	org := c.Organization()
-	if org != "" {
-		req.Header.Set("snyk-org-name", org)
-	}
-
-	req.Header.Set("snyk-request-id", requestId)
-	// https://www.keycdn.com/blog/http-cache-headers
-	req.Header.Set("Cache-Control", "private, max-age=0, no-cache")
+func (s *SnykCodeHTTPClient) encodeIfNeeded(method string, requestBody []byte) (*bytes.Buffer, error) {
+	b := new(bytes.Buffer)
+	mustBeEncoded := s.mustBeEncoded(method)
 	if mustBeEncoded {
-		req.Header.Set("Content-Type", "application/octet-stream")
-		req.Header.Set("Content-Encoding", "gzip")
-	} else {
-		req.Header.Set("Content-Type", "application/json")
-	}
-
-	log.Trace().Str("requestBody", string(requestBody)).Str("snyk-request-id", requestId).Msg("SEND TO REMOTE")
-	response, err := s.client().Do(req)
-	if err != nil {
-		log.Err(err).Str("method", method).Msgf("got http error")
-		s.errorReporter.CaptureErrorAndReportAsIssue(path, err)
-		return nil, err
-	}
-
-	defer func(Body io.ReadCloser) {
-		err := Body.Close()
+		enc := encoding.NewEncoder(b)
+		_, err := enc.Write(requestBody)
 		if err != nil {
-			log.Err(err).Msg("Couldn't close response body in call to Snyk Code")
+			return nil, err
 		}
-	}(response.Body)
-	responseBody, err := io.ReadAll(response.Body)
-	log.Trace().Str("response.Status", response.Status).Str("responseBody", string(responseBody)).Str("snyk-request-id",
-		requestId).Msg("RECEIVED FROM REMOTE")
-	if err != nil {
-		log.Err(err).Str("method", method).Msgf("error reading response body")
-		s.errorReporter.CaptureErrorAndReportAsIssue(path, err)
-		return nil, err
+	} else {
+		b = bytes.NewBuffer(requestBody)
 	}
+	return b, nil
+}
 
-	err = s.checkResponseCode(response)
-	if err != nil {
-		return nil, err
-	}
+func (s *SnykCodeHTTPClient) mustBeEncoded(method string) bool {
+	return method == http.MethodPost || method == http.MethodPut
+}
 
-	return responseBody, err
+var retryErrorCodes = map[int]bool{
```

Comment:
Yeah, can do.

Reply:
Looking at the error codes in CLI, I see different problems:

- they include client network stack connection errors (connection timeout, dns lookup, connection refused). These should IMHO never be retried
- e.g. 504 is not even in the list of http error codes, but that would have been one that would have been beneficial.

Reply:
So, I'll reduce the list to the following status codes. I won't do error parsing of the connection errors the http call may throw in this PR, as it would increase the scope and work quite a bit.

```
        http.StatusServiceUnavailable:  true,
	http.StatusBadGateway:          true,
	http.StatusGatewayTimeout:      true,
	http.StatusInternalServerError: true,
```

The retry loop was just an addition to the original missing files stuff that was the task in the user story, so please feel free to enhance it - or, if not acceptable, I can also remove it, whatever is better for you :).

Reply:
Thanks, makes sense.